runtime/compiler/runtime/JITServerAOTCache.cpp

```diff
@@ -967,6 +967,37 @@ JITServerAOTCache::getAOTHeaderRecord(const TR_AOTHeader *header, uint64_t clien
    return record;
    }
 
+const AOTCacheThunkRecord *
+JITServerAOTCache::getThunkRecord(const uint8_t *signature, uint32_t signatureSize)
+   {
+   OMR::CriticalSection cs(_thunkMonitor);
+
+   auto it = _thunkMap.find({ signature, signatureSize });
+   if (it != _thunkMap.end())
+      return it->second;
+
+   return NULL;
+   }
+
+const AOTCacheThunkRecord *
+JITServerAOTCache::createAndStoreThunk(const uint8_t *signature, uint32_t signatureSize, const uint8_t *thunkStart, uint32_t thunkSize)
+   {
```

Comment:
Shouldn't this be synchronized using the `_thunkMonitor`? Also, if I understand the code correctly, there is a possible race condition where a thunk record gets added after `getThunkRecord()` returns `NULL` but before the `createAndStoreThunk()` call, resulting in a memory leak for the newly allocated duplicate record. I think we need to check the `_thunkMap` again here before creating a new record.